api/v1alpha1/release_types_test.go

```diff
@@ -0,0 +1,671 @@
+	. "github.com/onsi/gomega/gstruct"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+const (
+	ReleaseReasonFake ReleaseReason = "Fake"
+	ReleaseReasonFoo  ReleaseReason = "Foo"
+	ReleaseReasonBar  ReleaseReason = "Bar"
+)
+
+type commonArgs struct {
+	status  metav1.ConditionStatus
+	reason  ReleaseReason
+	message string
+}
+
+var _ = Describe("release_types.go test collection", func() {
+
+	Context("ReleaseReason.String()", func() {
+		It("should return a string: fake", func() {
+			Expect(ReleaseReasonFake.String()).To(Equal("Fake"),
+				"Expected \"%+v\" to equal \"%+v\"", ReleaseReasonFake.String(), "Fake")
+		})
+
+		It("should return a string: foo", func() {
+			Expect(ReleaseReasonFoo.String()).To(Equal("Foo"),
+				"Expected \"%+v\" to equal \"%+v\"", ReleaseReasonFoo.String(), "Foo")
+		})
+
+		It("should return a string: bar", func() {
+			Expect(ReleaseReasonBar.String()).To(Equal("Bar"),
+				"Expected \"%+v\" to equal \"%+v\"", ReleaseReasonBar.String(), "Bar")
+		})
+
+		It("should return an empty string", func() {
+			var ReleaseEmptyReason ReleaseReason = ""
+			Expect(ReleaseEmptyReason.String()).To(Equal(""),
+				"Expected \"%+v\" to equal \"%+v\"", ReleaseEmptyReason.String(), "")
+		})
+	})
+
+	Context("HasBeenDeployed()", func() {
+		It("should return false because status.SnapshotEnvironmentBinding is nil", func() {
+			r := &Release{}
+			Expect(r.HasBeenDeployed()).To(BeFalse(),
+				"Expected to get: %+v ; Actually Got: %+v\n", false, true)
+		})
+
+		It("should return false because status.SnapshotEnvironmentBinding is empty", func() {
+			r := &Release{
+				Status: ReleaseStatus{
+					SnapshotEnvironmentBinding: "", // empty
+				},
+			}
+			Expect(r.HasBeenDeployed()).To(BeFalse(),
+				"Expected to get: %+v ; Actually Got: %+v\n", false, true)
+		})
+
+		It("should return true because status.SnapshotEnvironmentBinding is not empty", func() {
+			r := &Release{
+				Status: ReleaseStatus{
+					SnapshotEnvironmentBinding: "foo",
+				},
+			}
+			Expect(r.HasBeenDeployed()).To(BeTrue(),
+				"Expected to get: %+v ; Actually Got: %+v\n", true, false)
+		})
+	})
+
+	Context("HasStarted()", func() {
+		It("should return false because Status.startTime is nil", func() {
+			r := &Release{}
+			Expect(r.HasStarted()).To(BeFalse(),
+				"Expected to get: %+v ; Actually Got: %+v\n", false, true)
+		})
+
+		It("should return false because Status.startTime is zero", func() {
+			r := &Release{
+				Status: ReleaseStatus{
+					StartTime: &metav1.Time{}, // 0001-01-01 00:00:00 +0000 UTC
+					CompletionTime: &metav1.Time{
+						Time: time.Now(),
+					},
+					Conditions: []metav1.Condition{
+						{
+							Type:   releaseConditionType,
+							Status: metav1.ConditionTrue,
+						},
+					},
+				},
+			}
+			Expect(r.HasStarted()).To(BeFalse(),
+				"Expected to get: %+v ; Actually Got: %+v\n", false, true)
+		})
+
+		It("should return false because Status.startTime is nil", func() {
+			r := &Release{
```

Comment:
You could create a single Release in each context and modify the startTime in each `It` block. We don't like to influence other tests passing over the same objects, but within a `Context` is fine. You could even restore the default state at the end of each `It`. This would reduce the number of lines as this file has become quite big.

Reply:
This is probably outdated?
The structure is now setup with Ginkgo's `BeforeEach()` routine, and that dramatically reduced the number of lines. Therefore I'm not sure it's worth taking a step further to setup the structure in the context, and ensuring the structure is appropriate prior to each evaluation. However, I could be persuaded, for example if we ever wanted to optimize for a massive stress-test where we run up against garbage collection.